azafea/event_processors/endless/metrics/v3/model/_base.py

```diff
@@ -0,0 +1,513 @@
+    os_version = Column(Unicode, nullable=False)
+
+    @declared_attr
+    def channel_id(cls) -> Column:
+        return Column(Integer, ForeignKey(Channel.id), index=True)
+
+    @declared_attr
+    def channel(cls) -> relationship:
+        return relationship(Channel)
+
+    def __init__(self, payload: GLib.Variant, **kwargs: Dict[str, Any]) -> None:
+        payload_fields = self._parse_payload(payload)
+        fields = kwargs.copy()
+        fields.pop('singulars', None)
+        fields.pop('aggregates', None)
+        fields.update(payload_fields)
+
+        super().__init__(**fields)
+
+    def _parse_payload(self, maybe_payload: GLib.Variant) -> Dict[str, Any]:
+        payload = maybe_payload.get_maybe()
+
+        if self.__payload_type__ is None:
+            if payload is not None:
+                log.error('Metric event %s takes no payload, but got %s',
+                          self.__event_uuid__, payload)
+            return {}
+
+        if payload is None:
+            raise EmptyPayloadError(f'Metric event {self.__event_uuid__} needs a '
+                                    f'{self.__payload_type__} payload, but got none')
+
+        payload = get_variant(payload)
+        payload_type = payload.get_type_string()
+
+        if payload_type != self.__payload_type__:
+            raise WrongPayloadError(f'Metric event {self.__event_uuid__} needs a '
+                                    f'{self.__payload_type__} payload, but got '
+                                    f'{payload} ({payload_type})')
+
+        return self._get_fields_from_payload(payload)
+
+    @staticmethod
+    def _get_fields_from_payload(payload: GLib.Variant) -> Dict[str, Any]:
+        raise NotImplementedError('Implement this method in final event models')  # pragma: no cover
+
+
+class UnknownEvent(MetricEvent):
+    __abstract__ = True
+
+    event_id = Column(postgresql.UUID(as_uuid=True), nullable=False)
+    payload_data = Column(LargeBinary, nullable=False)
+
+    def _parse_payload(self, maybe_payload: GLib.Variant) -> Dict[str, Any]:
+        return {'payload_data': maybe_payload.get_data_as_bytes().get_data()}
+
+
+class InvalidEvent(UnknownEvent):
+    __abstract__ = True
+
+    error = Column(Unicode, nullable=False)
+
+
+class SingularEvent(MetricEvent):
+    __abstract__ = True
+
+    occured_at = Column(DateTime(timezone=True), nullable=False, index=True)
+
+
+class InvalidSingularEvent(SingularEvent, InvalidEvent):
+    __tablename__ = 'invalid_singular_event_v3'
+
+
+class UnknownSingularEvent(SingularEvent, UnknownEvent):
+    __tablename__ = 'unknown_singular_event_v3'
+
+
+class AggregateEvent(MetricEvent):
+    __abstract__ = True
+
+    period_start = Column(Date, nullable=False, index=True)
+    count = Column(BigInteger, nullable=False)
+
+
+class InvalidAggregateEvent(AggregateEvent, InvalidEvent):
+    __tablename__ = 'invalid_aggregate_event_v3'
+
+    receveid_period_start = Column(Unicode)
```

Comment:
"receveid" is a typo – but why is this field needed, can't we use `period_start` from the parent class?